File: backend/app/services/chat_service.py
import logging
import uuid
from datetime import datetime, timezone
from typing import List, Optional, Tuple

from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.conversation import Conversation, Message
from app.models.user import UserProfile
from app.services import llm_service, prompt_service

logger = logging.getLogger(__name__)


async def create_conversation(
    db: AsyncSession,
    user_profile_id: str,
    title: str = "新しい会話",
) -> Conversation:
    """Create a new conversation."""
    conv = Conversation(
        conversation_id=str(uuid.uuid4()),
        user_profile_id=user_profile_id,
        title=title,
    )
    db.add(conv)
    await db.commit()
    await db.refresh(conv)
    return conv


async def get_conversations(
    db: AsyncSession,
    user_profile_id: str,
) -> List[Conversation]:
    """Get all conversations for a user (without messages)."""
    result = await db.execute(
        select(Conversation)
        .where(Conversation.user_profile_id == user_profile_id)
        .order_by(Conversation.updated_at.desc())
    )
    return list(result.scalars().all())


async def get_conversation_detail(
    db: AsyncSession,
    conversation_id: str,
    user_profile_id: str,
) -> Optional[Conversation]:
    """Get a conversation with all messages."""
    result = await db.execute(
        select(Conversation)
        .options(selectinload(Conversation.messages))
        .where(
            Conversation.conversation_id == conversation_id,
            Conversation.user_profile_id == user_profile_id,
        )
    )
    return result.scalars().first()


async def get_message_count(
    db: AsyncSession,
    conversation_id: str,
) -> int:
    """Get message count for a conversation."""
    result = await db.execute(
        select(func.count(Message.message_id))
        .where(Message.conversation_id == conversation_id)
    )
    return result.scalar() or 0


async def add_message(
    db: AsyncSession,
    conversation_id: str,
    role: str,
    content: str,
    input_method: str = "text",
) -> Message:
    """Add a message to a conversation."""
    msg = Message(
        message_id=str(uuid.uuid4()),
        conversation_id=conversation_id,
        role=role,
        content=content,
        input_method=input_method,
    )
    db.add(msg)

    # Update conversation updated_at and title (if first user message)
    result = await db.execute(
        select(Conversation).where(Conversation.conversation_id == conversation_id)
    )
    conv = result.scalars().first()
    if conv and role == "user" and conv.title == "新しい会話":
        conv.title = content[:50] if len(content) > 50 else content

    await db.commit()
    await db.refresh(msg)
    return msg


async def send_chat_message(
    db: AsyncSession,
    user_profile_id: str,
    content: str,
    input_method: str = "text",
    conversation_id: Optional[str] = None,
) -> Tuple[Message, str, list]:
    """
    Process a chat message: save user message, generate AI response, save assistant message.
    Returns (assistant_message, conversation_id, sources).
    """
    # Create or get conversation
    if not conversation_id:
        conv = await create_conversation(db, user_profile_id, title=content[:50])
        conversation_id = conv.conversation_id

    # Save user message
    await add_message(db, conversation_id, "user", content, input_method)

    # Generate AI response
    ai_response = await _generate_ai_response(db, content, conversation_id)
    sources = ai_response.get("sources", [])

    # Save assistant message
    assistant_msg = await add_message(
        db, conversation_id, "assistant", ai_response["content"], "text"
    )

    return assistant_msg, conversation_id, sources


async def _generate_ai_response(
    db: AsyncSession,
    content: str,
    conversation_id: Optional[str] = None,
) -> dict:
    """
    Generate AI response using OpenAI GPT-4o-mini with admin prompt settings,
    few-shot examples, conversation history, and RAG context.
    Falls back to placeholder when OPENAI_API_KEY is not set.
    """
    # Get admin-configured system prompt and few-shot examples
    prompt_config = await prompt_service.get_current_prompt(db)
    system_prompt = prompt_config.system_prompt if prompt_config else ""
    few_shot_examples = prompt_config.few_shot_examples if prompt_config else []

    # Get conversation history (last 10 messages)
    chat_history = []
    if conversation_id:
        result = await db.execute(
            select(Message)
            .where(Message.conversation_id == conversation_id)
            .order_by(Message.created_at.desc())
            .limit(10)
        )
        messages = list(reversed(result.scalars().all()))
        chat_history = [
            {"role": msg.role, "content": msg.content}
            for msg in messages
        ]

    # RAG context (will be populated in Slice D)
    rag_context = None
    sources = []

    # Try vector search if available
    try:
        from app.services import vector_service
        search_results = await vector_service.search_similar(db, content, top_k=5)
        if search_results:
            rag_parts = []
            seen_titles = set()
            for i, chunk in enumerate(search_results, 1):
                title = chunk.get("document_title", "不明")
                rag_parts.append(f"[参考資料{i}: {title}]\n{chunk['content']}")
                if title not in seen_titles:
                    seen_titles.add(title)
                    sources.append({
                        "source_title": title,
                        "source_url": chunk.get("source_url"),
                        "source_type": chunk.get("source_type", "unknown"),
                    })
            rag_context = "\n\n".join(rag_parts)
    except ImportError:
        pass
    except Exception as e:
        logger.warning(f"RAG search failed, continuing without context: {e}")
        await db.rollback()

    # Call LLM
    llm_response = await llm_service.generate_chat_response(
        user_message=content,
        system_prompt=system_prompt,
        few_shot_examples=few_shot_examples,
        chat_history=chat_history,
        rag_context=rag_context,
    )

    # If no RAG sources and using placeholder, add default source
    if not sources:
        sources = [
            {
                "source_title": "社内マニュアル",
                "source_url": None,
                "source_type": "notion",
            }
        ]

    return {
        "content": llm_response["content"],
        "sources": sources,
    }

File: frontend/src/services/mock/chatService.ts
import type { Conversation, Message, ChatRequest, ChatResponse, KnowledgeReference } from '@/types/index.ts';
import { MOCK_CONVERSATIONS } from './data/conversations.ts';

const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

let conversations = [...MOCK_CONVERSATIONS];

interface MockResponseEntry {
  content: string;
  sources: KnowledgeReference[];
}

const MOCK_RESPONSES: MockResponseEntry[] = [
  {
    content: 'お疲れさまです。そのご質問について確認しますね。\n\n社内マニュアルによると、該当の手続きは総務部で受け付けています。まずは社内ポータルの「各種申請」ページをご覧いただくのが早いかと思います。\n\n他にご不明な点があれば、お気軽にお聞きください。',
    sources: [
      { source_title: '社内マニュアル - 各種手続き', source_url: null, source_type: 'notion' },
    ],
  },
  {
    content: 'なるほど、それは気になりますよね。\n\n確認したところ、その件については研修資料の第3章に記載があるようです。もしお手元に資料がない場合は、研修チームに連絡すると共有してもらえますよ。\n\nまた何かあればいつでもどうぞ。',
    sources: [
      { source_title: '新人研修ガイド 第3章', source_url: null, source_type: 'google_drive' },
    ],
  },
  {
    content: 'ご質問ありがとうございます。\n\nその手続きについては、直属の上長に確認いただくのが確実です。一般的な流れとしては、まず社内システムから申請を出し、上長の承認後に処理されます。\n\n詳しい手順が必要でしたら、もう少し具体的にお聞きしますね。',
    sources: [
      { source_title: '就業規則 第5章', source_url: null, source_type: 'notion' },
      { source_title: '社内申請フロー図', source_url: null, source_type: 'google_drive' },
    ],
  },
];

// @MOCK_TO_API: GET {API_PATHS.CHAT.CONVERSATIONS}
export const mockChatService = {
  async getConversations(): Promise<Conversation[]> {
    await delay(300);
    return conversations.map(({ messages, ...rest }) => ({
      ...rest,
      message_count: messages?.length ?? 0,
    }));
  },

  // @MOCK_TO_API: GET {API_PATHS.CHAT.CONVERSATION_DETAIL}
  async getConversation(conversationId: string): Promise<Conversation | null> {
    await delay(300);
    return conversations.find(c => c.conversation_id === conversationId) ?? null;
  },

  // @MOCK_TO_API: POST {API_PATHS.CHAT.SEND}
  async sendMessage(request: ChatRequest): Promise<ChatResponse> {
    await delay(1500);

    const now = new Date().toISOString();
    const responseIndex = Math.floor(Math.random() * MOCK_RESPONSES.length);
    const mockResponse = MOCK_RESPONSES[responseIndex];

    const userMessage: Message = {
      message_id: `msg-${Date.now()}`,
      role: 'user',
      content: request.content,
      input_method: request.input_method,
      created_at: now,
    };

    const assistantMessage: Message = {
      message_id: `msg-${Date.now() + 1}`,
      role: 'assistant',
      content: mockResponse?.content ?? '',
      input_method: 'text',
      created_at: new Date(Date.now() + 2000).toISOString(),
    };

    let conversationId = request.conversation_id;

    if (!conversationId) {
      conversationId = `conv-${Date.now()}`;
      const newConv: Conversation = {
        conversation_id: conversationId,
        title: request.content.slice(0, 30) + (request.content.length > 30 ? '...' : ''),
        created_at: now,
        updated_at: now,
        messages: [userMessage, assistantMessage],
      };
      conversations = [newConv, ...conversations];
    } else {
      const conv = conversations.find(c => c.conversation_id === conversationId);
      if (conv) {
        if (!conv.messages) conv.messages = [];
        conv.messages.push(userMessage, assistantMessage);
        conv.updated_at = now;
      }
    }

    return {
      message: assistantMessage,
      conversation_id: conversationId,
      sources: mockResponse?.sources,
    };
  },

  // @MOCK_TO_API: POST {API_PATHS.CHAT.NEW_CONVERSATION}
  async createConversation(): Promise<Conversation> {
    await delay(300);
    const conv: Conversation = {
      conversation_id: `conv-${Date.now()}`,
      title: '新しい会話',
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
      messages: [],
    };
    conversations = [conv, ...conversations];
    return conv;
  },
};
